mentat/code_context.py

```diff
@@ -199,147 +117,306 @@ async def refresh(self, replace_paths: bool | None = False):
+
+        def _feature_relative_path(f: CodeFile) -> str:
+            return os.path.relpath(f.path, self.config.git_root)
+
+        return sorted(include_features, key=_feature_relative_path)
+
+    async def _get_auto_features(
+        self,
+        code_file_manager: "CodeFileManager",
+        parser: "Parser",
+        model: str,
+        include_features: list[CodeFile],
+        max_tokens: Optional[int] = None,
+    ) -> list[CodeFile]:
+        # Generate all possible permutations for all files in the project.
+        candidate_features = list[CodeFile]()
+        for path in get_non_gitignored_files(self.config.git_root):
+            if not path.is_dir() and not is_file_text_encoded(path):
+                continue
+            permutations: list[tuple[CodeMessageLevel, str | None]] = [
+                (CodeMessageLevel.CODE, None),
+                (CodeMessageLevel.FILE_NAME, None),
+            ]
+            if not self.settings.no_code_map:
+                permutations += [
+                    (CodeMessageLevel.CMAP_FULL, None),
+                    (CodeMessageLevel.CMAP, None),
+                ]
+            if self.diff_context.target and path in self.diff_context.files:
+                permutations += [
+                    (level, self.diff_context.target) for level, _ in permutations
                 ]
-                message = "\n".join(message)
-                await stream.send(message, color="yellow")
+            for level, diff in permutations:
+                feature = CodeFile(path, level=level, diff=diff)
+                candidate_features.append(feature)
 
-        return "\n".join(code_message)
+        # Sort candidates by relevance/density.
+        tokens_remaining = 1e6 if max_tokens is None else max_tokens
 
-    async def add_file(self, code_file: CodeFile):
-        stream = SESSION_STREAM.get()
+        async def _calculate_feature_score(feature: CodeFile) -> float:
+            score = 0.0
+            tokens = await feature.count_tokens(
+                self.config, code_file_manager, parser, model
+            )
+            if tokens == 0:
+                raise MentatError(f"Feature {feature} has 0 tokens.")
+            if feature.diff is not None:
+                score += 1
+            if feature.level == CodeMessageLevel.FILE_NAME:
+                score += 0.1
+            if feature.level == CodeMessageLevel.CMAP:
+                score += 0.25
+            elif feature.level == CodeMessageLevel.CMAP_FULL:
+                score += 0.5
+            elif feature.level == CodeMessageLevel.CODE:
+                score += 0.75
+            score /= tokens
+            return score
+
+        all_features = include_features.copy()
+        candidate_scores_task = [
+            _calculate_feature_score(f) for f in candidate_features
+        ]
+        candidate_scores = await asyncio.gather(
+            *candidate_scores_task, return_exceptions=True
+        )
```

Comment:
When I run this PR on Mentat it crashes with many of these tasks returning `OSError(24, 'Too many open files')`

We should check to make sure we're properly closing files in `CodeFileManager`, but assuming there's no issue with that maybe we can add a Semaphore to `_calculate_feature_score`? I haven't used them before but here's an example from GPT-4:

```sem = asyncio.Semaphore(10)  # Adjust value to suit your needs

async def process_file(filename):
    async with sem:
        with open(filename, 'r') as f:
            # process file

await asyncio.gather(*(process_file(f) for f in files))
```

Reply:
Small world gpt just recommended adding a semaphore to llm_api to limit concurrent api calls.

Reply:
TIL about semaphores. Let me know if you still see the issue with the new commit @biobootloader.

Reply:
What about a global semaphore as a ContextVar?

Reply:
> What about a global semaphore as a ContextVar?

We'd only want it to apply to file openings, right? So maybe it would be used just in CodeFileManager?

Reply:
> Let me know if you still see the issue with the new commit @biobootloader.

seems fixed!